### ISSUE TITLE
[question] How to set dropout.training=False in CustomPolicy when loading model or testing

### ISSUE THREAD

**Issue by zkx741481546** [hill-a/stable-baselines#833]
Q1:
I have seen the 'CustomPolicy' part in the doc but found no detail about dropout.
Here is my Policy
`import gym
import tensorflow as tf
from stable_baselines.common.policies import ActorCriticPolicy, register_policy, nature_cnn, MlpPolicy, \
    FeedForwardPolicy
from stable_baselines.common.vec_env import DummyVecEnv
from stable_baselines import A2C
from Config import seed

class CustomPolicy(ActorCriticPolicy):
    def __init__(self, sess, ob_space, ac_space, n_env, n_steps, n_batch, reuse=False, **kwargs):
        super(CustomPolicy, self).__init__(sess, ob_space, ac_space, n_env, n_steps, n_batch, reuse=reuse, scale=True)

        with tf.variable_scope("model", reuse=reuse):
            activ = tf.nn.relu
            net_arch = [dict(vf=[256, 128, 64, 32], pi=[256, 128, 64, 32])]
            l2_scale = 0.01
            dropout_rate = 0.0
            for k, v in kwargs.items():
                if k == 'act_fun':
                    activ = v
                elif k == 'net_arch':
                    net_arch = v
                elif k == 'l2_scale':
                    l2_scale = v
                elif k == 'dropout_rate':
                    dropout_rate = v
                else:
                    raise ("不支持的Policy_arg:{}".format(k))
            # extracted_features = nature_cnn(self.processed_obs, **kwargs)
            extracted_features = self.processed_obs
            if len(ob_space.shape) > 1:
                extracted_features = tf.layers.flatten(extracted_features)
            index_code = 0
            arch_pi_and_vf = None
            for arch in net_arch:
                if isinstance(arch, int):
                    extracted_features = tf.layers.dropout(activ(
                        tf.layers.dense(extracted_features, arch, name='feature_extract' + str(arch),
                                        kernel_initializer=tf.contrib.layers.l2_regularizer(l2_scale))),
                        rate=dropout_rate,
                        seed=seed,
                        training=True)
                    index_code += 1
                elif isinstance(arch, dict):
                    arch_pi_and_vf = arch
                else:
                    raise ("自定义网络参数不合法: " + str(arch))
            pi_layer_size = arch_pi_and_vf['pi']
            vf_layer_size = arch_pi_and_vf['vf']
            pi_h = extracted_features
            for i, layer_size in enumerate(pi_layer_size):
                pi_h = tf.layers.dropout(activ(tf.layers.dense(pi_h, layer_size, name='pi_fc' + str(i),
                                                               kernel_regularizer=tf.contrib.layers.l2_regularizer(
                                                                   l2_scale))), rate=dropout_rate, seed=seed, training=True)
            pi_latent = pi_h

            vf_h = extracted_features
            for i, layer_size in enumerate(vf_layer_size):
                vf_h = tf.layers.dropout(activ(tf.layers.dense(vf_h, layer_size, name='vf_fc' + str(i),
                                                               kernel_regularizer=tf.contrib.layers.l2_regularizer(
                                                                   l2_scale))), rate=dropout_rate, seed=seed, training=True)
            value_fn = tf.layers.dense(vf_h, 1, name='vf')
            vf_latent = vf_h

            self._proba_distribution, self._policy, self.q_value = \
                self.pdtype.proba_distribution_from_latent(pi_latent, vf_latent, init_scale=0.01)

        self._value_fn = value_fn
        self._setup_init()

    def step(self, obs, state=None, mask=None, deterministic=False):
        if deterministic:
            action, value, neglogp = self.sess.run([self.deterministic_action, self.value_flat, self.neglogp],
                                                   {self.obs_ph: obs})
        else:
            action, value, neglogp = self.sess.run([self.action, self.value_flat, self.neglogp],
                                                   {self.obs_ph: obs})
        return action, value, self.initial_state, neglogp

    def proba_step(self, obs, state=None, mask=None):
        return self.sess.run(self.policy_proba, {self.obs_ph: obs})

    def value(self, obs, state=None, mask=None):
        return self.sess.run(self.value_flat, {self.obs_ph: obs})
`

Is there any way to change dropout(training=False) after the model was loaded?

Q2:
And I'm using Stable-Baselines 2.10.0 with my CustomPolicyNetwork, there is no error or warning when I loading the model without customObject, such as policy and policy_kwargs. And the network works normally.
Is that means in the 2.10.0 version, no more customObject is needed when loading?

**Comment by Miffyli** [comment]
Q1: Looking at [this StackOverflow](https://stackoverflow.com/questions/44971349/how-to-turn-off-dropout-for-testing-in-tensorflow), you have to create additional placeholders to easily disable/enable dropout. I suggest you try this.

Q2: I am not sure what the issue was before, but if all the custom objects are serializeble (pickle-able), then loading agents should work without providing additional items.

**Comment by zkx741481546** [comment]
> 问题 1：查看[此堆栈溢出](https://stackoverflow.com/questions/44971349/how-to-turn-off-dropout-for-testing-in-tensorflow)，您必须创建其他占位符才能轻松禁用/启用退位。我建议你试试这个
> 
> 问题 2：我不确定之前的问题是什么，但如果所有自定义对象都是序列化的（可拾取），则加载代理应该在不提供其他项的情况下工作。
> 
> 如果没有与稳定基线相关的其他问题，则可以关闭此问题。

Thanks for your reply. I have seen that StackOverFlow question, but I'm not sure how to combine a placeholder with Stable-Baselines framework, if I pass the training=False argument through policy_kwargs, there will be a error cause of the different between my policy_kwargs and the policy_kwargs in the data file.
I just saw the source code in TRPO.load(), and wrote this function to change the policy_kwargs before setup_model. There is no error or warning when I use this, but I'm not sure if it's right.
`policy_args = dict(act_fun=gelu, net_arch=[dict(vf=[256, 128, 64, 32], pi=[256, 128, 64, 32])], l2_scale=0.01, dropout_rate=0.5)
`
`def LoadCustomPolicyForTest(model_path):
    data, params = TRPO._load_from_file(model_path)
    # 设置dropout比率为0.，模型内部会自动设置training为False
    data['policy_kwargs']['dropout_rate'] = 0.
    model = TRPO(policy=data["policy"], env=None, _init_setup_model=False)
    model.__dict__.update(data)
    model.setup_model()
    model.load_parameters(params)
    return model
`
I changed my policy into this, which can auto set the training argument into False when dropout_rate == 0.
`class CustomPolicy(ActorCriticPolicy):
    def __init__(self, sess, ob_space, ac_space, n_env, n_steps, n_batch, reuse=False, **kwargs):
        super(CustomPolicy, self).__init__(sess, ob_space, ac_space, n_env, n_steps, n_batch, reuse=reuse, scale=True)

        with tf.variable_scope("model", reuse=reuse):
            activ = tf.nn.relu
            net_arch = [dict(vf=[256, 128, 64, 32], pi=[256, 128, 64, 32])]
            l2_scale = 0.01
            dropout_rate = 0.0
            training = False
            for k, v in kwargs.items():
                if k == 'act_fun':
                    activ = v
                elif k == 'net_arch':
                    net_arch = v
                elif k == 'l2_scale':
                    l2_scale = v
                elif k == 'dropout_rate':
                    dropout_rate = v
                    if dropout_rate == 0.:
                        training = False
                    else:
                        training = True
                else:
                    raise ("不支持的Policy_arg:{}".format(k))
            # extracted_features = nature_cnn(self.processed_obs, **kwargs)
            extracted_features = self.processed_obs
            if len(ob_space.shape) > 1:
                extracted_features = tf.layers.flatten(extracted_features)
            index_code = 0
            arch_pi_and_vf = None
            for arch in net_arch:
                if isinstance(arch, int):
                    extracted_features = tf.layers.dropout(activ(
                        tf.layers.dense(extracted_features, arch, name='feature_extract' + str(arch),
                                        kernel_initializer=tf.contrib.layers.l2_regularizer(l2_scale))),
                        rate=dropout_rate,
                        seed=seed,
                        training=training)
                    index_code += 1
                elif isinstance(arch, dict):
                    arch_pi_and_vf = arch
                else:
                    raise ("自定义网络参数不合法: " + str(arch))
            pi_layer_size = arch_pi_and_vf['pi']
            vf_layer_size = arch_pi_and_vf['vf']
            pi_h = extracted_features
            for i, layer_size in enumerate(pi_layer_size):
                pi_h = tf.layers.dropout(activ(tf.layers.dense(pi_h, layer_size, name='pi_fc' + str(i),
                                                               kernel_regularizer=tf.contrib.layers.l2_regularizer(
                                                                   l2_scale))), rate=dropout_rate, seed=seed,
                                         training=training)
            pi_latent = pi_h

            vf_h = extracted_features
            for i, layer_size in enumerate(vf_layer_size):
                vf_h = tf.layers.dropout(activ(tf.layers.dense(vf_h, layer_size, name='vf_fc' + str(i),
                                                               kernel_regularizer=tf.contrib.layers.l2_regularizer(
                                                                   l2_scale))), rate=dropout_rate, seed=seed,
                                         training=training)
            value_fn = tf.layers.dense(vf_h, 1, name='vf')
            vf_latent = vf_h

            self._proba_distribution, self._policy, self.q_value = \
                self.pdtype.proba_distribution_from_latent(pi_latent, vf_latent, init_scale=0.01)

        self._value_fn = value_fn
        self._setup_init()

    def step(self, obs, state=None, mask=None, deterministic=False):
        if deterministic:
            action, value, neglogp = self.sess.run([self.deterministic_action, self.value_flat, self.neglogp],
                                                   {self.obs_ph: obs})
        else:
            action, value, neglogp = self.sess.run([self.action, self.value_flat, self.neglogp],
                                                   {self.obs_ph: obs})
        return action, value, self.initial_state, neglogp

    def proba_step(self, obs, state=None, mask=None):
        return self.sess.run(self.policy_proba, {self.obs_ph: obs})

    def value(self, obs, state=None, mask=None):
        return self.sess.run(self.value_flat, {self.obs_ph: obs})`
And I use it like this:
`model = LoadCustomPolicyForTest(model_path)`

**Comment by Miffyli** [comment]
On a quick glimpse this should work. Sadly there is no proper test for this to see if it works, and you have to study mid-network activations and such to see if things are working correctly. 

You can close this issue if there are no further questions.

**Comment by zkx741481546** [comment]
> 快速一瞥，这应该工作。可悲的是，没有适当的测试，看看它是否工作，你必须研究中间网络激活，如，看看事情是否正常工作。
> 
> 如果没有其他问题，您可以关闭此问题。

Thank you!
